File: cs/thread_03.c
#include <stdio.h>
#include <unistd.h>
#include <pthread.h>

void * foo() {
    printf("excuted foo()!!!\n");
    return NULL;
}

void * bar() {
    printf("excuted bar()!!!\n");
    return NULL;
}

void * baz() {
    printf("excuted baz()!!!\n");
    return NULL;
}

// 각기 다른 작업을 하는 스레드 만들기

int main()
{
    pthread_t thread1;
    pthread_t thread2;
    pthread_t thread3;

    pthread_create(&thread1, NULL, foo, NULL);
    pthread_create(&thread2, NULL, bar, NULL);
    pthread_create(&thread3, NULL, baz, NULL);

    pthread_join(thread1, NULL);
    pthread_join(thread2, NULL);
    pthread_join(thread3, NULL);

    return 0;
}

File: cs/thread_03.py
import threading
import os

def foo():
    print("I'm foo!!!")
    print("thread id", threading.get_native_id())
    print("process id", os.getpid())
    print("*" * 50)
    
def bar():
    print("I'm bar!!!")
    print("thread id", threading.get_native_id())
    print("process id", os.getpid())
    print("*" * 50)

def baz():
    print("I'm baz!!!")
    print("thread id", threading.get_native_id())
    print("process id", os.getpid())
    print("*" * 50)

if __name__ == '__main__':
    print("process id", os.getpid())
    thread1 = threading.Thread(target=foo).start()
    thread2 = threading.Thread(target=bar).start()
    thread3 = threading.Thread(target=baz).start()

# 세 개의 스레드 모두 각기 다른 스레드지만, 이들은 모두 동일한 프로세스를 공유하고 있다.